ethcore/src/blockchain/blockchain.rs

```diff
@@ -1350,11 +1359,13 @@ impl BlockChain {
 				}
 			},
 			BlockLocation::BranchBecomingCanonChain(ref data) => {
-				let ancestor_number = self.block_number(&data.ancestor).unwrap();
+				let ancestor_number = self.block_number(&data.ancestor)
+					.expect("Inserted block's ancestor number always exists; qed");
```

Comment:
"hash belongs to an ancestor of an inserted block; ancestors of an inserted block are always available; block number of an inserted block is always available; qed"

Actually while writing this I realised that this could fail because the second condition doesn't always hold true (because of warp sync) right?

Reply:
Ah looks like that's indeed. This function is called by `insert_unordered_block` which can get unordered?

Reply:
It seems that this is fine, although all the logic is not really apparent (so we may consider to do some refactorings for this): In wrap sync, the location is always set to `BlockLocation::CanonChain` so we will never reach this match statement.

Reply:
Thanks for looking this up. Yes, it's not clear from just looking at this code that invariant holds. But it makes sense that we do consider the warped block to be canon (what else can we do if we already trusted the warp?).